In [10]:
#  Ebnable HTML/CSS 
from IPython.core.display import HTML
HTML("<link href='https://fonts.googleapis.com/css?family=Passion+One' rel='stylesheet' type='text/css'><style>div.attn { font-family: 'Helvetica Neue'; font-size: 30px; line-height: 40px; color: #FFFFFF; text-align: center; margin: 30px 0; border-width: 10px 0; border-style: solid; border-color: #5AAAAA; padding: 30px 0; background-color: #DDDDFF; }hr { border: 0; background-color: #ffffff; border-top: 1px solid black; }hr.major { border-top: 10px solid #5AAA5A; }hr.minor { border: none; background-color: #ffffff; border-top: 5px dotted #CC3333; }div.bubble { width: 65%; padding: 20px; background: #DDDDDD; border-radius: 15px; margin: 0 auto; font-style: italic; color: #f00; }em { color: #AAA; }div.c1{visibility:hidden;margin:0;height:0;}div.note{color:red;}</style>")

___
**GROUP 2** 
- Name 1: Taylor Bonar
- Name 2: Robert Burigo
- Name 3: Rashmi Patel
- Name 4: Scott Englerth
________

# In Class Assignment Three
In the following assignment you will be asked to fill in python code and derivations for a number of different problems. Please read all instructions carefully and turn in the rendered notebook (or HTML of the rendered notebook)  before the end of class.

<a id="top"></a>
## Contents
* <a href="#Loading">Loading the Data</a>
* <a href="#distance">Measuring Distances</a>
* <a href="#KNN">K-Nearest Neighbors</a>
* <a href="#naive">Naive Bayes</a>

________________________________________________________________________________________________________
<a id="Loading"></a>
<a href="#top">Back to Top</a>
## Downloading the Document Data
Please run the following code to read in the "20 newsgroups" dataset from sklearn's data loading module.

In [2]:
from __future__ import print_function

In [3]:
from sklearn.datasets import fetch_20newsgroups_vectorized
import numpy as np

# this takes about 30 seconds to compute, read the next section while this downloads
ds = fetch_20newsgroups_vectorized(subset='train')

# this holds the continuous feature data (which is tfidf)
print('features shape:', ds.data.shape) # there are ~11000 instances and ~130k features per instance
print('target shape:', ds.target.shape) 
print('range of target:', np.min(ds.target),np.max(ds.target))
print('Data type is', type(ds.data), float(ds.data.nnz)/(ds.data.shape[0]*ds.data.shape[1])*100, '% of the data is non-zero')

features shape: (11314, 130107)
target shape: (11314,)
range of target: 0 19
Data type is <class 'scipy.sparse.csr.csr_matrix'> 0.1214353154362896 % of the data is non-zero


## Understanding the Dataset
Look at the description for the 20 newsgroups dataset at http://qwone.com/~jason/20Newsgroups/. You have just downloaded the "vectorized" version of the dataset, which means all the words inside the articles have gone through a transformation that binned them into 130 thousand features related to the words in them.  

**Question Set 1**:
- How many instances are in the dataset? 
- What does each instance represent? 
- How many classes are in the dataset and what does each class represent?
- Would you expect a classifier trained on this data would generalize to documents written in the past week? Why or why not?
- Is the data represented as a sparse or dense matrix?

___
- How many instances are in the dataset? **11314 instances**
- What does each instance represent? **A different newsgroup where each newsgroup corresponds to a different topic where all the words in the article have been transformed/vectorized into 130,000 features**
- How many classes are in the dataset and what does each class represent? **There are 20 classes representing a newsgroup with each corresponding to a different topic**
- Would you expect a classifier trained on this data would generalize to documents written in the past week? Why or why not? **We think the data could possibly generalize documents to certain newsgroups, but due to the older data, we think the accuracy could be reduced**
- Is the data represented as a sparse or dense matrix? **This is a sparse index as the data has 0.1214353154362896 % of non-zero values**

___
<a id="distance"></a>
<a href="#top">Back to Top</a>
## Measures of Distance
In the following block of code, we isolate three instances from the dataset. The instance "`a`" is from the group *computer graphics*, "`b`" is from from the group *recreation autos*, and "`c`" is from group *recreation motorcycle*. **Exercise for part 2**: Calculate the: 
- (1) Euclidean distance
- (2) Cosine distance 
- (3) Jaccard similarity 

between each pair of instances using the imported functions below. Remember that the Jaccard similarity is only for binary valued vectors, so convert vectors to binary using a threshold. 

**Question for part 2**: Which distance seems more appropriate to use for this data? **Why**?

In [9]:
from scipy.spatial.distance import cosine
from scipy.spatial.distance import euclidean
from scipy.spatial.distance import jaccard
import numpy as np

# get first instance (comp)
idx = 550  # Index of comp.graphics
a = ds.data[idx].todense()  # returns "dense" (most values are non-zero) matrix
a_class = ds.target_names[ds.target[idx]]  # Retrieves the index/instance's target name
print('Instance A is from class', a_class)

# get second instance (autos)
idx = 4000
b = ds.data[idx].todense()
b_class = ds.target_names[ds.target[idx]]
print('Instance B is from class', b_class)

# get third instance (motorcycle)
idx = 7000
c = ds.data[idx].todense()
c_class = ds.target_names[ds.target[idx]]
print('Instance C is from class', c_class)

ab_euclidean = euclidean(a, b)
ab_cosine = cosine(a, b)
ab_jaccard = jaccard(a, b)

ac_euclidean = euclidean(a, c)
ac_cosine = cosine(a, c)
ac_jaccard = jaccard(a, c)

bc_euclidean = euclidean(b,c)
bc_cosine = cosine(b,c)
bc_jaccard = jaccard(b,c)
print(f'a={a}')
print(f'b={b}')
print(f'c={c}')

# Enter distance comparison below for each pair of vectors:
p = 'Placeholder'
print(f'\n\nEuclidean Distance\n\tab:{ab_euclidean}\n\tac:{ac_euclidean}\n\tbc:{bc_euclidean}')
print(f'Cosine Distance\n\tab:{ab_cosine} \n\tac:{ac_cosine} \n\tbc:{bc_cosine}')
print(f'Jaccard Dissimilarity (vectors should be boolean values)\n\tab:{ab_jaccard}' \
f'\n\tac:{ac_jaccard}\n\tbc:{bc_jaccard}')

print('\n\nThe most appropriate distance is Cosine')

Instance A is from class comp.graphics
Instance B is from class rec.autos
Instance C is from class rec.motorcycles
a=[[0. 0. 0. ... 0. 0. 0.]]
b=[[0. 0. 0. ... 0. 0. 0.]]
c=[[0. 0. 0. ... 0. 0. 0.]]


Euclidean Distance
	ab:1.098518467187086
	ac:1.1891405425398234
	bc:0.9177794226661622
Cosine Distance
	ab:0.6033714113755322 
	ac:0.7070276149559529 
	bc:0.4211595343347173
Jaccard Dissimilarity (vectors should be boolean values)
	ab:1.0
	ac:1.0
	bc:1.0


The most appropriate distance is Cosine


___
# Start of Live Session Assignment
___
<a id="KNN"></a>
<a href="#top">Back to Top</a>
## Using scikit-learn with KNN
Now let's use stratified cross validation with a holdout set to train a KNN model in `scikit-learn`. Use the example below to train a KNN classifier. The documentation for `KNeighborsClassifier` is here: http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsClassifier.html  

**Exercise for part 3**: Use the code below to test what value of `n_neighbors` works best for the given data. *Note: do NOT change the metric to be anything other than `'euclidean'`. Other distance functions are not optimized for the amount of data we are working with.* 

**Question for part 3**: What is the accuracy of the best classifier you can create for this data (by changing only the `n_neighbors` parameter)? 

In [27]:
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
from IPython.html import widgets 

cv = StratifiedShuffleSplit(n_splits=1, test_size = 0.5, train_size=0.5)

# fill in the training and testing data and save as separate variables
for trainidx, testidx in cv.split(ds.data,ds.target):
    # note that these are sparse matrices
    X_train = ds.data[trainidx] 
    X_test = ds.data[testidx] 
    y_train = ds.target[trainidx]
    y_test = ds.target[testidx]

# fill in your code  here to train and test
# calculate the accuracy and print it for various values of K
K = 5 
clf = KNeighborsClassifier(n_neighbors=K, weights='uniform', metric='euclidean')

fit = clf.fit(X_train,y_train)

predictions = clf.predict(X_test)
print(f'Predictions List: {predictions}')

print(clf.score(X_test, y_test))
acc = accuracy_score(y_test, predictions)
print(f'Accuracy = {acc}')

#=====================================
best_k=0
best_acc=0

for k in range(1,10):
    if(k % 2 is not 0):
        print(f'\n\nTesting with k={k}')
        clf = KNeighborsClassifier(n_neighbors=k, weights='uniform', metric='euclidean')
        fit = clf.fit(X_train,y_train)

        predictions = clf.predict(X_test)
        print(f'Predictions List: {predictions}')

        print(clf.score(X_test, y_test))
        acc = accuracy_score(y_test, predictions)
        print(f'Accuracy = {acc}')
        
        if best_acc < acc:
            best_acc = acc
            best_k = k
        
        
print('\n\nBest Accuracy of classifier with %d neighbors is: %.2f'%(best_k,best_acc))


Predictions List: [ 1  5 11 ... 18  1  0]
0.48842142478345413
Accuracy = 0.48842142478345413


Testing with k=1
Predictions List: [ 1 12 11 ... 18  1  0]
0.6056213540745978
Accuracy = 0.6056213540745978


Testing with k=3
Predictions List: [ 1  5 11 ... 18  1 18]
0.521831359377762
Accuracy = 0.521831359377762


Testing with k=5
Predictions List: [ 1  5 11 ... 18  1  0]
0.48842142478345413
Accuracy = 0.48842142478345413


Testing with k=7
Predictions List: [ 1  1  0 ... 18  1 18]
0.4718048435566555
Accuracy = 0.4718048435566555


Testing with k=9
Predictions List: [ 1  1 11 ... 18  1 18]
0.4587237051440693
Accuracy = 0.4587237051440693


Best Accuracy of classifier with 1 neighbors is: 0.61


**Question for part 3**: With sparse data, does the use of a KDTree representation make sense? Why or Why not?

Our instincts for using KDTree representation with sparse matrices/data is not ideal as our original dataset has about 130,000 features resulting in high dimensionality which can result in potential fruitless splits in our KDTree. 

_____
## KNN extensions - Centroids
Now lets look at a very closely related classifier to KNN, called nearest centroid. In this classifier (which is more appropriate for big data scenarios and sparse data), the training step is used to calculate the centroids for each class. These centroids are saved. Unknown attributes, at prediction time, only need to have distances calculated for each saved centroid, drastically decreasing the time required for a prediction. 

**Exercise for part 4**: Use the template code below to create a nearest centroid classifier. Test which metric has the best cross validated performance: Euclidean, Cosine, or Manhattan. In `scikit-learn` you can see the documentation for NearestCentroid here: 
- http://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestCentroid.html#sklearn.neighbors.NearestCentroid

and for supported distance metrics here:
- http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.distance_metrics.html#sklearn.metrics.pairwise.distance_metrics

In [32]:
from sklearn.neighbors import NearestCentroid

# the parameters for the nearest centroid metric to test are:
#    l1 (manhanta, l2, and cosine (all are optimized)
clf = NearestCentroid(metric='euclidean')

# fill in your code here
e_clf = NearestCentroid(metric='l2')
e_fit = e_clf.fit(X_train,y_train)
e_predictions = e_clf.predict(X_test)
e_acc = e_clf.score(X_test,y_test)

print(f'euclidean centroid accuracy = {e_acc}')


c_clf = NearestCentroid(metric='cosine')
c_fit = c_clf.fit(X_train,y_train)
c_predictions = c_clf.predict(X_test)
c_acc = c_clf.score(X_test,y_test)

print(f'cosine centroid accuracy = {c_acc}')

m_clf = NearestCentroid(metric='l1')
m_fit = m_clf.fit(X_train,y_train)
m_predictions = m_clf.predict(X_test)
m_acc = m_clf.score(X_test,y_test)

print(f'manhanttan centroid accuracy = {m_acc}')

print('The best distance metric is cosine centroid accuracy')

C:\Users\Taylo\anaconda3\envs\ML1\lib\site-packages\sklearn\neighbors\_nearest_centroid.py:150: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "
C:\Users\Taylo\anaconda3\envs\ML1\lib\site-packages\sklearn\neighbors\_nearest_centroid.py:150: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "


euclidean centroid accuracy = 0.4138235814035708
cosine centroid accuracy = 0.4843556655471098


C:\Users\Taylo\anaconda3\envs\ML1\lib\site-packages\sklearn\neighbors\_nearest_centroid.py:150: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "


manhanttan centroid accuracy = 0.3063461198515114
The best distance metric is cosine centroid accuracy


___
<a id="naive"></a>
<a href="#top">Back to Top</a>
## Naive Bayes Classification
Now let's look at the use of the Naive Bayes classifier. The 20 newsgroups dataset has 20 classes and about 130,000 features per instance. Recall that the Naive Bayes classifer calculates a posterior distribution for each possible class. Each posterior distribution is a multiplication of many conditional distributions: 

$${\arg \max}_{j} \left(p(class=j)\prod_{i} p(attribute=i|class=j) \right)$$

where $p(class=j)$ is the prior and $p(attribute=i|class=j)$ is the conditional probability.

**Question for part 5**: With this many classes and features, how many different conditional probabilities need to be parameterized? How many priors need to be parameterized?

**20 priors and way too many conditional probabilities**

In [35]:
probabilities=20*130107
print(f'{probabilities} conditional probablities')

2602140 conditional probablities


___
## Naive Bayes in Scikit-learn
Scikit has several implementations of the Naive Bayes classifier: `GaussianNB`, `MultinomialNB`, and `BernoulliNB`. Look at the documentation here: http://scikit-learn.org/stable/modules/naive_bayes.html Take a look at each implementation and then answer this question: 

**Questions for part 6**: 
- If the instances contain mostly continuous attributes, would it be better to use Gaussian Naive Bayes, Multinomial Naive Bayes, or Bernoulli? And Why? 
- What if the data is sparse, does this change your answer? Why or Why not?

6.1: Gaussian Naive Bayes would be the ideal implementation as it is usually chosen to represent the class-conditional probability for continuous attributes. 

6.2: No, because the data is still continuous, but you may want to consider multinomial naive bayes for comparisons. 
___

## Naive Bayes Comparison
For the final section of this notebook let's compare the performance of Naive Bayes for document classification. Look at the parameters for `MultinomialNB`, and `BernoulliNB` (especially `alpha` and `binarize`). 

**Exercise for part 7**: Using the example code below, change the parameters for each classifier and see how accurate you can make the classifiers on the test set. 

**Question for part 7**: Why are these implementations so fast to train? What does the `'alpha'` value control in these models (*i.e.*, how does it change the parameterizations)? 

In [49]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.naive_bayes import BernoulliNB


clf_mnb = MultinomialNB(alpha=1.0)
clf_bnb = BernoulliNB(alpha=1.0, binarize=0.0)

# fill in your code here
mnb_fit = c_clf.fit(X_train,y_train)
mnb_predictions = c_clf.predict(X_test)
mnb_acc = c_clf.score(X_test,y_test)

print(f'Multinomial NB accuracy: {mnb_acc}')

bnb_fit = c_clf.fit(X_train,y_train)
bnb_predictions = c_clf.predict(X_test)
bnb_acc = c_clf.score(X_test,y_test)

print(f'Bernoulli NB accuracy: {bnb_acc}')

best_mnb_acc=0
best_mnb_alpha=0
best_bnb_alpha=0
best_bnb_acc=0

for a in np.linspace(0,1,11):
    print(f'Testing with alpha set to {a}')
    clf_mnb = MultinomialNB(alpha=a)
    clf_bnb = BernoulliNB(alpha=a, binarize=0.0)

    mnb_fit = clf_mnb.fit(X_train,y_train)
    mnb_predictions = clf_mnb.predict(X_test)
    mnb_acc = clf_mnb.score(X_test,y_test)
    
    bnb_fit = clf_bnb.fit(X_train,y_train)
    bnb_predictions = clf_bnb.predict(X_test)
    bnb_acc = clf_bnb.score(X_test,y_test)
    
    if best_bnb_acc < bnb_acc:
        best_bnb_acc = bnb_acc
        best_bnb_alpha = a
        
    if best_mnb_acc < mnb_acc:
        best_mnb_acc = mnb_acc
        best_mnb_alpha = a
        
print(f'The best alpha for Multinomial NB is {best_mnb_alpha} with an accuracy of {best_mnb_acc}')
print(f'The best alpha for Bernoulli NB is {best_bnb_alpha} with an accuracy of {best_bnb_acc}')


print('These classifiers are so fast because we have sparse data')
print('The alpha values controls the additive smoothing parameter (accounts for features not present in the learning samples) which causes the classification to skew towards the class with more records resulting in underfitting')


C:\Users\Taylo\anaconda3\envs\ML1\lib\site-packages\sklearn\neighbors\_nearest_centroid.py:150: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "
C:\Users\Taylo\anaconda3\envs\ML1\lib\site-packages\sklearn\neighbors\_nearest_centroid.py:150: UserWarning: Averaging for metrics other than euclidean and manhattan not supported. The average is set to be the mean.
  warnings.warn("Averaging for metrics other than "


Multinomial NB accuracy: 0.4843556655471098
Bernoulli NB accuracy: 0.4843556655471098
Testing with alpha set to 0.0


C:\Users\Taylo\anaconda3\envs\ML1\lib\site-packages\sklearn\naive_bayes.py:509: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)
C:\Users\Taylo\anaconda3\envs\ML1\lib\site-packages\sklearn\naive_bayes.py:509: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10
  'setting alpha = %.1e' % _ALPHA_MIN)


Testing with alpha set to 0.1
Testing with alpha set to 0.2
Testing with alpha set to 0.30000000000000004
Testing with alpha set to 0.4
Testing with alpha set to 0.5
Testing with alpha set to 0.6000000000000001
Testing with alpha set to 0.7000000000000001
Testing with alpha set to 0.8
Testing with alpha set to 0.9
Testing with alpha set to 1.0
The best alpha for Multinomial NB is 0.0 with an accuracy of 0.8652996287785045
The best alpha for Bernoulli NB is 0.0 with an accuracy of 0.8516881739437865
These classifiers are so fast because we have sparse data
The alpha values controls the additive smoothing parameter (accounts for features not present in the learning samples) which causes the classification to skew towards the class with more records resulting in underfitting


________________________________________________________________________________________________________

That's all! Please **upload your rendered notebook to blackboard** and please include **team member names** in the notebook submission.